### ISSUE TITLE
[Win32] gmake removes perly.c when building 5.39.5

### ISSUE THREAD

**Issue by sisyphus** [Perl/perl5#21651]
This one feels like a pebcak, but I haven't so far been able to nail it.
No such problem with 5.39.4 and earlier.
For simplicity, I've created an entirely new folder (C:\New) into which I unpack perl-5.39.5.tar.gz.
Then 'cd' to C:\New\perl-5.39.5\win32 and run <code>gmake CCHOME=C:\winlibs-ucrt-1320\mingw64</code>
Perl builds fine, using the default GNUmakefile values, but the process terminates with:
```
...
gmake[1]: Leaving directory 'C:/New/perl-5.39.5/cpan/podlators'
Running pm_to_blib for cpan/version directly
rm ..\perly.c
```
Any ideas as to what the cause might be ?
If I don't provide the "CCHOME" argument, the build throws an error ... but still manages to <code>rm ..\perly.c</code> before terminating.

If I want to then run <code>gmake test</code> I first have to reinstall perly.c - which again needs to be reinstalled before running <code>gmake install</code>.

As a check, I've also built a freshly unpacked 5.39.4 source using the same toolset and the same gmake command. That works fine and perly.c does not get clobbered. The process terminates correctly with:
```
gmake[1]: Leaving directory 'C:/New/perl-5.39.4/cpan/podlators'
Running pm_to_blib for cpan/version directly
```
I'll have another crack at it tomorrow. In the meantime, any ideas are welcome.

Here's the <code>perl -V</code> output  for the perl-5.39.5 build:
```
C:\New\perl-5.39.5\win32>..\perl -I..\lib -V
Summary of my perl5 (revision 5 version 39 subversion 5) configuration:

  Platform:
    osname=MSWin32
    osvers=10.0.22621.2715
    archname=MSWin32-x64-multi-thread
    uname=''
    config_args='undef'
    hint=recommended
    useposix=true
    d_sigaction=undef
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=undef
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='gcc'
    ccflags =' -DWIN32 -DWIN64 -DPERL_TEXTMODE_SCRIPTS -DMULTIPLICITY -DPERL_IMPLICIT_SYS -DUSE_PERLIO -D__USE_MINGW_ANSI_STDIO -fwrapv -fno-strict-aliasing -mms-bitfields'
    optimize='-Os'
    cppflags='-DWIN32'
    ccversion=''
    gccversion='13.2.0'
    gccosandvers=''
    intsize=4
    longsize=4
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='long long'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='g++'
    ldflags ='-s -L"c:\perl\lib\CORE" -L"C:\winlibs-ucrt-1320\mingw64\lib" -L"C:\winlibs-ucrt-1320\mingw64\x86_64-w64-mingw32\lib" -L"C:\winlibs-ucrt-1320\mingw64\lib\gcc\x86_64-w64-mingw32\13.2.0"'
    libpth=C:\winlibs-ucrt-1320\mingw64\lib C:\winlibs-ucrt-1320\mingw64\x86_64-w64-mingw32\lib C:\winlibs-ucrt-1320\mingw64\lib\gcc\x86_64-w64-mingw32\13.2.0
    libs= -lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32 -ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
    perllibs= -lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32 -ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
    libc=
    so=dll
    useshrplib=true
    libperl=libperl539.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_win32.xs
    dlext=dll
    d_dlsymun=undef
    ccdlflags=' '
    cccdlflags=' '
    lddlflags='-shared -s -L"c:\perl\lib\CORE" -L"C:\winlibs-ucrt-1320\mingw64\lib" -L"C:\winlibs-ucrt-1320\mingw64\x86_64-w64-mingw32\lib" -L"C:\winlibs-ucrt-1320\mingw64\lib\gcc\x86_64-w64-mingw32\13.2.0"'


Characteristics of this binary (from libperl):
  Compile-time options:
    HAS_LONG_DOUBLE
    HAS_TIMES
    HAVE_INTERP_INTERN
    MULTIPLICITY
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_HASH_FUNC_SIPHASH13
    PERL_HASH_USE_SBOX32
    PERL_IMPLICIT_SYS
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_USE_SAFE_PUTENV
    USE_64_BIT_INT
    USE_ITHREADS
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
    USE_THREAD_SAFE_LOCALE
  Built under MSWin32
  Compiled at Nov 21 2023 00:18:53
  @INC:
    ..\lib
    C:/New/perl-5.39.5/lib
```

**Comment by tonycoz** [comment]
Is there an attempt (and failure) to run `yacc` or `bison` earlier in the output?

If I extract the archive and look at the modification dates/times:
```
tony@GANYMEDE /cygdrive/c/Users/Tony/dev/perl
$ tar xzf perl-5.39.5.tar.gz

tony@GANYMEDE /cygdrive/c/Users/Tony/dev/perl
$ perl -MTime::HiRes=stat -E 'say "$_: ", (stat $_)[9] for @ARGV' perl-5.39.5/perly.[yc]
perl-5.39.5/perly.c: 1676917329
perl-5.39.5/perly.y: 1692478271
```
it looks like `perly.y` is newer than `perly.c`, and it may be the default rules are trying to rebuild `perly.c`, failing, and then make is trying to remove the possibly incorrect `perly.c`.

I don't see that behaviour building here though, the only references to `perly.c` in the build log are the two compilations, one for miniperl and the other for the main binary.

**Comment by sisyphus** [comment]
No mention of "yacc" or "bison" anywhere in the build, and I see only the same 2 references to "perly.c" during the build.
The problem with perly.c seems to be avoided if I use make-4.2.1 (which I was using until a week or so ago) instead of make-4.4.1.

But now there's another annoying issue in that <code>make distclean</code> is not cleaning things up like it should, and I have to start from a freshly unpacked source for every mingw-w64 build I do.
When I try to re-use a distcleaned source, the "make" step bails out at:
```
make[1]: Leaving directory 'D:/comp-1320-ucrt/perl-5.39.5/ext/Sys-Hostname'
Makefile.PL: The "xdefine" exists, skipping the configure step.
Use "D:\comp-1320-ucrt\perl-5.39.5\miniperl.exe Makefile.PL --configure"
or: "D:\comp-1320-ucrt\perl-5.39.5\miniperl.exe Makefile.PL --force
to force the configure step.
Warning: No Makefile!
make[1]: Entering directory 'D:/comp-1320-ucrt/perl-5.39.5/dist/Time-HiRes'
make[1]: *** No rule to make target 'all'.  Stop.
make[1]: Leaving directory 'D:/comp-1320-ucrt/perl-5.39.5/dist/Time-HiRes'
make[1]: Entering directory 'D:/comp-1320-ucrt/perl-5.39.5/dist/Time-HiRes'
make[1]: *** No rule to make target 'all'.  Stop.
make[1]: Leaving directory 'D:/comp-1320-ucrt/perl-5.39.5/dist/Time-HiRes'
Unsuccessful make(dist/Time-HiRes): code=512 at ..\make_ext.pl line 584.
make: *** [GNUmakefile:1560: Extensions] Error 2
```
There could still be a pebcak element at play here.
It might take me a while to sort through it all.

**Comment by sisyphus** [comment]
I've done some more fiddling about, and established that the removal of <code>perly.c</code> happens only when I use make-4.4.1 (which comes from my winlibs gcc-13.2.0 installation).
Using make-4.2.1 (from whereabouts unknown) avoids this issue.

I've only been able to intermittently reproduce the malfunctioning <code>make distclean</code>, but I think it doesn't happen if we stick with the default configuration of  <code>PLMAKE=gmake</code>.

If anyone sees a good reason to keep this issue open, please speak up. Otherwise I'll close it in 2 days time (Friday night local time).

**Comment by sisyphus** [comment]
Re-opening because I've finally reached an understanding of how to reliably tickle (and work around) the bug.

Firstly, this is an issue that affects only perl-5.39.5 and later. There's no such issue with perl-5.39.4 and earlier.

For me, <code>perly.c</code> gets removed by gmake at the conclusion of  the <code>gmake</code> (and <code>gmake test</code> and <code>gmake install</code>) steps only if I use the gmake-4.4.1 that ships with Strawberrty Perl 5.38.0. (That is the only gmake-4.4.1 that I've tried).
If I use gmake-4.2.1 (origin unknown), there's no problem.

Picking up on @tonycoz's one liner, in perl-5.39.4, I find that both <code>perly.y</code> and <code>perly.c</code> have the same timestamp:
```
Owner@DESKTOP-88J497T /cygdrive/d/comp-1310
$ perl -MTime::HiRes=stat -E 'say "$_: ", (stat $_)[9] for @ARGV' perl-5.39.5/perly.[yc]
perl-5.39.4/perly.c: 1697885259
perl-5.39.4/perly.y: 1697885259
```
As Tony showed, that's not the case with 5.39.5:
```
Owner@DESKTOP-88J497T /cygdrive/d/comp-1310
$ perl -MTime::HiRes=stat -E 'say "$_: ", (stat $_)[9] for @ARGV' perl-5.39.5/perly.[yc]
perl-5.39.5/perly.c: 1676917329
perl-5.39.5/perly.y: 1692478271
```
Turns out that all I apparently have to do is to <code>touch perl-5.39.5/perly.c</code> prior to building, and make-4.4.1 happily builds perl-5.39.5, and leaves <code>perly.c</code> in place.

Is it good enough to simply fix the timestamp in the perl  source distro (and if so, how ?), or is make-4.4.1 being buggy ?

**Comment by tonycoz** [comment]
Could you please provide a build log from the build where this happens?

**Comment by sisyphus** [comment]
Build log for failing perl-5.39.5 is attached as 441.txt.

Firstly, I unpacked fresh 5.39.5 source into C:\comp-1310\perl-5.39.5.441.
I then logged the building of that source, using gmake (version 4.4.1) and gcc-13.1.0, as provided by the ("portable") StrawberryPerl-5.38.0.
It ran successfully, but removed perly.c.
Upon completion, I renamed (moved) C:\comp-1310\perl-5.39.5.441 to C:\comp-1310\perl-5.39.5.441_first.

I then again unpacked fresh 5.39.5 source into a newly re-created C:\comp-1310\perl-5.39.5.441 and repeated the first build, but this time using gmake-4.2.1.
It ran successfully, and did not remove perly.c

I also logged that second build - attached as 421.txt.

Ignoring whitespace differences between the 2 builds, the <code>diff -wu</code> shows only 4 lines of difference between 441.txt and 421.txt
One of those lines is, of course <code>rm ..\perly.c</code>, and I haven't yet been able to detect the difference in the other 3 lines.
I've attached that diff, as diff.txt.

Just yell out if you want a log of a build done using gmake-4.4.1, having first touched perly.c.

UPDATE: I've done <code>diff -u</code> between 441.txt and the gmake-4.4.1 build log of the source containing the touched perly.c.
It shows that the only difference between those 2 logs is the <code>rm ..\perly.c</code> that appears only in 441,txt.

[441.txt](https://github.com/Perl/perl5/files/13519040/441.txt)
[421.txt](https://github.com/Perl/perl5/files/13519042/421.txt)
[diff.txt](https://github.com/Perl/perl5/files/13519052/diff.txt)
